File: ZAJECIA_5/Sortowanie_topologiczne.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

void sortowanie_topologiczne(int n, vector<vector<int>> &graf)
{
    vector<int> stopien_wejsciowy(n, 0);
    for (int i = 0; i < n; i++)
    {
        for (int sasiad : graf[i])
        {
            stopien_wejsciowy[sasiad]++;
        }
    }
    cout << "Stopnie wejsciowe:" << endl;
    for (int i = 0; i < stopien_wejsciowy.size(); i++)
    {
        cout << stopien_wejsciowy[i] << " ";
    }
    cout << endl;
    queue<int> kolejka;
    for (int i = 0; i < n; i++)
    {
        if (stopien_wejsciowy[i] == 0)
        {
            kolejka.push(i);
        }
    }
    vector<int> wynik;
    while (!kolejka.empty())
    {
        int v = kolejka.front();
        kolejka.pop();
        wynik.push_back(v);

        for (int sasiad : graf[v])
        {
            stopien_wejsciowy[sasiad]--;
            if (stopien_wejsciowy[sasiad] == 0)
            {
                kolejka.push(sasiad);
            }
        }
    }
    cout << "Sortowanie topologiczne:" << endl;
    for (int v : wynik)
    {
        cout << v << " ";
    }
    cout << endl;
}

int main()
{
    int n = 7;
    // Przyklad z wykladu wziety
    vector<vector<int>> graf = {{4, 2}, {0, 4}, {}, {1}, {2}, {1, 4}, {2}};
    sortowanie_topologiczne(n, graf);
    return 0;
}

File: ZAJECIA_5/Sortowanie_topologiczne.js
function sortowanieTopologiczne(n, graf) {
    let stopienWejsciowy = new Array(n).fill(0);
    for (let listaSasiadow of graf) {
        for (let sasiad of listaSasiadow) {
            stopienWejsciowy[sasiad]++;
        }
    }
    console.log("Stopnie wejsciowe:");
    console.log(stopienWejsciowy.join(" "));
    let kolejka = [];
    for (let i = 0; i < n; i++) {
        if (stopienWejsciowy[i] == 0) {
            kolejka.push(i);
        }
    }
    let wynik = [];
    while (kolejka.length > 0) {
        let v = kolejka.shift();
        wynik.push(v);
        for (let sasiad of graf[v]) {
            stopienWejsciowy[sasiad]--;
            if (stopienWejsciowy[sasiad] == 0) {
                kolejka.push(sasiad);
            }
        }
    }
    console.log("Sortowanie topologiczne:")
    console.log(wynik.join(" "));
}

let n = 7;
let graf = [[4, 2], [0, 4], [], [1], [2], [1, 4], [2]];
sortowanieTopologiczne(n, graf);